File: 2016/day09.rb
def decompress(input)
  output = ''

  until input.empty?
    if input[0] == '('
      ending = input.index(')')
      a, b = input[1..ending-1].join.split('x').map(&:to_i)

      input = input.drop(ending + 1)
      output += (input[0...a] * b).join

      input = input.drop(a)

    else
      output += input.shift
    end
  end

  output.size
end

def decompress2(input)
  return input.size unless input.index('(')

  sum = 0

  until input.empty?
    if input[0] == '('
      ending = input.index(')')
      a, b = input[1..ending-1].join.split('x').map(&:to_i)

      input = input.drop(ending + 1)
      sum += decompress2(input[0...a]) * b

      input = input.drop(a)

    else
      input.shift
      sum += 1
    end
  end

  sum
end

s = STDIN.read.strip

puts decompress(s.chars)
puts decompress2(s.chars)

File: 2016/day09.py
data = input().strip()

def decompress(data):
    count = 0

    while '(' in data:
        s = data.index('(')
        e = data.index(')') + 1
        a, b = map(int, data[s+1:e-1].split('x'))
        count += s + a * b
        data = data[(e+a):]

    return count

def decompress2(data):
    if '(' not in data:
        return len(data)

    count = 0

    while '(' in data:
        s = data.index('(')
        e = data.index(')') + 1
        a, b = map(int, data[s+1:e-1].split('x'))
        count += s + decompress2(data[e:e+a]) * b
        data = data[(e+a):]

    return count

print(decompress(data))
print(decompress2(data))
